add alb support, upgrade elb resources and documentation

PixarV opened the pull request:
FEATURES:

* **New Resource:** `aws_lb_listener_certificate`
* **New Resource:** `aws_lb_listener_rule`

ENHANCEMENTS / BUG FIXES:

* resource/aws_lb:
  *  update valid values for `load_balancer_type`, remove default value
  *  allow updating `subnets` and `subnet_mapping` arguments for ALB and NLB (only adding new subnets)
  * `subnets` and `subnet_mapping` arguments can no longer be specified together
* resource/aws_listener:
  * `default_action`
    * the block can only be specified once
    * `forward` and `target_group_arn` arguments can no longer be specified together
    * mark `target_group_arn` as deprecated:  it's recommended to use `forward` block instead
    * update valid values for `type`
    * introduce new argument: `order`
    * change max `forward.weight` value to 256
  * make `port` and `protocol` arguments required
  * update valid values for `protocol`
* resource/aws_target_group:
  * `health_check`
    * remove TCP default values for `healthy_threshold`, `protocol`, `unhealthy_threshold` arguments
    * update valid values for `interval`, `port`, `protocol` arguments
  * make `port`,  `protocol`, `vpc_id` arguments required
  * update valid values for `protocol`, `protocol_version`, `target_type` arguments
  * forbid updating `protocol_version`
* resource/aws_target_group_attachment:
  * set max and min `port` value

DOCUMENTATION FIXES:

* Upgrade documentation for the section **ELB (Elastic Load Balancing)**

Commit:
provider: add new elb resources

Commit:
website: aws_lb_listener_certificate support

Commit:
elbv2,website: aws_listener_rule support

Commit:
elbv2,website: upgrade aws_lb + ds and documentation

Parameters update:
- remove custom diff that ignores subnets changes for nlb
- load_balancer_type: update possible values and remove default value
- subnets, subnet_mapping: exactly one of them must be specified in
configuration
- subnet_mapping: support update

Commit:
elbv2,website: upgrade aws_listener + ds and documentation

Parameters update:
- default_action: exactly one block per configuration
- default_action.forward, default_action.target_group_arn: cannot be
specified together
- default_action.type: update possible values
- default_action.order: support the parameter
- default_action.forward.weight: update max value
- port: optional -> required
- protocol: optional -> required, update possible values

Commit:
elbv2,website: upgrade aws_target_group + ds and documentation

Parameters update:
- remove custom diff that checks some combinations for parameters: it
shows false positive errors for computed parameters
- health_check:
-- healthy_threshold, protocol, unhealthy_threshold: remove default
TCP based values, add computed because api sets default values for them
-- healthy_threshold, unhealthy_threshold, matcher, path: check if empty
 before send parameter to api
-- interval: set range for possible values
-- port: update possible values
-- protocol: update possible values
- port: optional -> required
- protocol: optional -> required, update possible values
- protocol_version: mark as ForceNew, update possible values
- target_type: update possible values
- vpc_id: optional -> required

Commit:
elbv2,website: upgrade aws_target_group_attachment + documentation

Parameters update:
- port: add validation for values

Commit:
mkdocs: add new elb resources

DiGa66 commented on website/docs/d/lb_listener.md:
```diff
@@ -43,10 +37,58 @@ data "aws_lb_listener" "selected443" {
+* `certificate_arn` - The Amazon Resource Name (ARN) of the IAM server certificate.
+* `default_action` - The default action that is applied to incoming requests.
+  The structure of this block is [described below](#default_action).
+* `id` - he Amazon Resource Name (ARN) of the listener.
```
```suggestion
* `id` - The Amazon Resource Name (ARN) of the listener.
```
тут все правильно? не ID?

PixarV replied:
Да, все корректно. арн сохраняется в `id`, тк именно он используется для запросов к апи.

cactusFriday replied:
he -> The gentle ping)

PixarV replied:
Накопировала..

DiGa66 commented on website/docs/d/lb_listener.md:
```diff
@@ -43,10 +37,58 @@ data "aws_lb_listener" "selected443" {
+
+The `target_group` block has the following structure:
+
+* `arn` - The Amazon Resource Name (ARN) of the target group to forward traffic.
```
```suggestion
* `arn` - The Amazon Resource Name (ARN) of the target group to forward traffic to.
```
речь же о целевой группе, на которую трафик направляется?

PixarV replied:
Да) поправила

DiGa66 commented on website/docs/d/lb_listener.md:
```diff
@@ -43,10 +37,58 @@ data "aws_lb_listener" "selected443" {
+
+The `forward` block has the following structure:
+
+* `target_group` - List of target groups to forward traffic.
```
```suggestion
* `target_group` - List of target groups to forward traffic to.
```
см. коммент ниже

DiGa66 commented on website/docs/d/lb_target_group.md:
```diff
@@ -3,48 +3,65 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+* `health_check` - The health check configuration.
+  The structure of this block is [described below](#health_check).
+* `id` - The Amazon Resource Name (ARN) of the target group.
```
Не ID?

PixarV replied:
Нет, там действительно арн

DiGa66 commented on website/docs/d/lb_target_group.md:
```diff
@@ -3,48 +3,65 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+The `health_check` block has the following structure:
+
+* `enabled` - Indicates whether health check will be enabled.
```
это же датасорс, по идее атрибут уже задан
```suggestion
* `enabled` - Indicates whether health check is enabled.
```

DiGa66 commented on website/docs/d/lb_target_group.md:
```diff
@@ -3,48 +3,65 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `path` - The destination for the health check request.
+* `port` - The port used to perform health checks on targets.
+* `protocol` - The protocol used to perform health checks on targets.
+* `timeout` - The amount of time, in seconds, during which no response means a failed health check.
```
```suggestion
* `timeout` - The amount of time, in seconds, after which no response indicates a failed health check.
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `load_balancer_type` - (Optional) The type of the load balancer.
+    * _Valid values_: `application`, `network`
+* `name` - (Optional) The name of the load balancer.
+    * _Length of the value_: From 1 to 32 symbols
```
в пользовательской доке это "Length constraints"
если твой вариант, то тогда просто "Value length"

PixarV replied:
Поменяла везде на `Value length`

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+        * `name` cannot be specified with `name_prefix`
+        * The value can contain only Latin letters, numbers and hyphens (`-`)
+        * The value must start and end with a Latin letter or number
+        * The value cannot start with prefix `internal-`
```
```suggestion
        * The value cannot start with the prefix `internal-`
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `name_prefix` - (Optional) Creates a unique name beginning with the specified prefix.
+    * _Length of the value_: From 1 to 6 symbols
+    * _Constraints_:
+        * `name_prefix` cannot be specified with `name`
```
не очень понял, что тут имеется ввиду
cannot be specified if `name` is set (?)

PixarV replied:
Поправила формулировку

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+        * `name_prefix` cannot be specified with `name`
+        * The value constraints are the same as `name`
+
+-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with a prefix `tf-lb`.
```
```suggestion
-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with the prefix `tf-lb`.
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with a prefix `tf-lb`.
+
+* `subnet_mapping` - (Optional) List of subnet ID to IP address mappings.
```
```suggestion
* `subnet_mapping` - (Optional) List of subnet-ID-to-IP-address mappings.
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+* `subnet_mapping` - (Optional) List of subnet ID to IP address mappings.
+  The structure of this block is [described below](#subnet_mapping).
+    * _Constraints_: `subnet_mapping` is required if `subnets` is not specified
```
```suggestion
    * _Constraints_: `subnet_mapping` is required if the `subnets` attribute is not specified
```
а то немного странно выглядит мн.ч. с глаголом в ед.ч.

PixarV replied:
Поправила, только вместо атрибута аргумент

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Constraints_: `subnet_mapping` is required if `subnets` is not specified
+* `subnets` - (Optional, Editable) List of subnet IDs.
+    * _Constraints_:
+        * `subnets` is required if `subnet_mapping` is not specified
```
```suggestion
        * The `subnets` attribute is required if `subnet_mapping` is not specified
```

PixarV replied:
атрибут -> аргумент

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `subnets` - (Optional, Editable) List of subnet IDs.
+    * _Constraints_:
+        * `subnets` is required if `subnet_mapping` is not specified
+        * All subnets must be from different available zones
```
```suggestion
        * All subnets must be from different availability zones
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+        * `subnets` is required if `subnet_mapping` is not specified
+        * All subnets must be from different available zones
+
+~> **Note** You can only add new subnets to the `subnets` list, removing subnets is not available.
```
```suggestion
~> **Note** You can only add new subnets to the `subnets` list, subnets cannot be removed.
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+* `subnet_id` - (Required) The ID of the subnet.
+* `allocation_id` - (Optional) The ID of the Elastic IP address allocation.
+  The **internet-facing** load balancer will be available at this IP address.
```
решила болдом выделять? не тяжело будет выглядеть?

PixarV replied:
Заменила болд на курсив

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `private_ipv4_address` - (Optional) The private IP address within the specified subnet.
+  The **internal** load balancer will be available at this IP address.
+
+~> **Note** All subnets specified in the `subnet_mapping` blocks must be from different available zones.
```
```suggestion
~> **Note** All subnets specified in the `subnet_mapping` blocks must be from different availability zones.
```

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
-* `tags_all` - A map of tags assigned to the resource, including those inherited from the provider [`default_tags` configuration block][default-tags].
-* `zone_id` - The canonical hosted zone ID of the load balancer (to be used in a Route 53 Alias record).
-* `subnet_mapping.*.outpost_id` - ID of the Outpost containing the load balancer.
+* `id` - The Amazon Resource Name (ARN) of the load balancer.
```
ID ?

PixarV replied:
Если бы..)

DiGa66 commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Constraints_: `forward` can be specified only if `type` is `forward` and `target_group_arn` is not specified
+* `order` - (Optional, Editable) The sequential number of the action.
+    * _Valid values_: From 1 to 50000
+* `target_group_arn` - (Optional, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic.
```
```suggestion
* `target_group_arn` - (Optional, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic to.
```

DiGa66 commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
-The following arguments are optional:
-
-* `stickiness` - (Optional) Configuration block for target group stickiness for the rule. Detailed below.
+* `target_group` - (Required, Editable) List of target groups to forward traffic.
```
```suggestion
* `target_group` - (Required, Editable) List of target groups to forward traffic to.
```

DiGa66 commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
-* `stickiness` - (Optional) Configuration block for target group stickiness for the rule. Detailed below.
+* `target_group` - (Required, Editable) List of target groups to forward traffic.
+  The structure of this block is [described below](#target_group).
+    * _Size of the list_: From 1 to 5 elements
```
```suggestion
    * _List size_: From 1 to 5 elements
```
либо Length constraints

PixarV replied:
Поменяла везде на `List size`

DiGa66 commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
+The `target_group` block has the following structure:
 
-The following arguments are required:
+* `arn` - (Required, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic.
```
```suggestion
* `arn` - (Required, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic to.
```

DiGa66 commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
 
-The following arguments are optional:
+* `arn` - The Amazon Resource Name (ARN) of the listener.
+* `id` - he Amazon Resource Name (ARN) of the listener.
```
```suggestion
* `id` - The Amazon Resource Name (ARN) of the listener.
```
Два атрибута с одним значением, но разными названиями?

PixarV replied:
Да, все так. `id` - это служебный параметр, который должен быть у каждого ресурса. Значение из этого параметра используется в запросах к апи для идентификации ресурса. В случае с elb api для идентификации ресруса используются арны.

DiGa66 commented on website/docs/r/lb_listener_certificate.md:
```diff
@@ -0,0 +1,106 @@
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - `listener_arn` and `certificate_arn` separated by an underscore (`_`).
```
странный разделитель

PixarV replied:
Думаю, он выбран потому что символы типа `-` могут использоваться в арнах

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+* `listener_arn` - (Required) The Amazon Resource Name (ARN) of the listener.
+    * _ARN Format_: `arn:c2:elasticloadbalancing::<project-name>@<customer-name>:listener/<app|net>/lb-12345678/li-12345678`
+* `priority` - (Optional) Priority of the rule.
+  If `priority` is not set, it will receive the value of next highest available priority.
```
```suggestion
  If `priority` is not set, its value will be set to the next highest available priority.
```

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+* `redirect` - (Optional, Editable) The block with information about redirecting requests to another URL.
+  The structure of this block is [described below](#redirect).
+    * _Constraints_: `redirect` can be specified only if `type` is `redirect`
+* `target_group_arn` - (Optional, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic.
```
```suggestion
* `target_group_arn` - (Optional, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic to.
```

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+* `status_code` - (Required, Editable) The HTTP code of the response.
+    * _Valid values_: `2XX`, `4XX`, `5XX`, where `X` is a digit
+* `message_body` - (Optional, Editable) The message of the response.
+    * _Length of the value_: From 0 to 1024 symbols
```
см. аналог. прим.

PixarV replied:
Теперь `Value length`

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+The `forward` block has the following structure:
+
+* `target_group` - (Required, Editable) List of target groups to forward traffic.
```
```suggestion
* `target_group` - (Required, Editable) List of target groups to forward traffic to.
```

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+* `target_group` - (Required, Editable) List of target groups to forward traffic.
+  The structure of this block is [described below](#target_group).
+    * _Size of the list_: From 2 to 5 elements
```
см. аналог. прим.

PixarV replied:
Теперь `List size`

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+The `target_group` block has the following structure:
+
+* `arn` - (Required, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic.
```
```suggestion
* `arn` - (Required, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic to.
```

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+    * _Length of the value_: From 3 to 255 symbols
+    * _Constraints_: See all the constraints in [ELB API documentation][redirect-action-config]
+* `path` - (Optional, Editable) The case-sensitive absolute path.
+  The value can contain `#{host}`, `#{path}`, and `#{port}`.
```
либо везде ставим зпт перед and, либо нигде (см. например стр. 273)
и так и так правильно, но вообще рекомендуют ставить, чтобы не возникало ситуаций как в русском "казнить нельзя помиловать" (хотя в англ. это не так критично)

PixarV replied:
Давай будем ставить) поправила в других местах тоже

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+  The value can contain `#{host}`, `#{path}`, and `#{port}`.
+  This component is not percent-encoded.
+    * _Default value_: `/#{path}`
+    * _Length of the value_: From 1 to 128 symbols
```
см. анал. ком.

PixarV replied:
Теперь `Value length`

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+    * _Constraints_: See all the constraints in [ELB API documentation][redirect-action-config]
+* `path` - (Optional, Editable) The case-sensitive absolute path.
+  The value can contain `#{host}`, `#{path}`, and `#{port}`.
+  This component is not percent-encoded.
```
или cannot be (?)

PixarV replied:
Тут имеется ввиду, что эта часть урла будет использоваться как есть без дополнительного кодирования

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+The `condition` block has the following structure:
+
+* `host_header` - (Optional, Editable) The block with information about the host headers the host name in the URL should match.
```
тоже лучше добавить that как в стр. 306

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+The `host_header` block has the following structure:
+
+* `values` - (Required, Editable) List of the domain host names that the host name in the URL will be compared to.
```
а не host domain names
домен хоста ведь, не хост домена (?)

PixarV replied:
Я взяла этот вариант из [пользовательской документации](https://docs.k2.cloud/en/api/elb/datatypes/HostHeaderConditionConfig.html#hostheaderconditionconfig). В оригинале доменные имена хостов

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+* `values` - (Required, Editable) List of the domain host names that the host name in the URL will be compared to.
+  If multiple names are specified, a logical OR operator is applied for the condition.
+    * _Length of the value_: From 3 to 255 symbols
```
см. анал. ком

PixarV replied:
Теперь `Value length`

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+* `values` - (Required, Editable) List of the path patterns that the path in the URL will be compared to.
+  If multiple values are specified, a logical OR operator is applied for condition.
+    * _Length of the value_: From 1 to 128 symbols
```
см.

PixarV replied:
Теперь `Value length`

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+The `path_pattern` block has the following structure:
+
+* `values` - (Required, Editable) List of the path patterns that the path in the URL will be compared to.
+  If multiple values are specified, a logical OR operator is applied for condition.
```
```suggestion
  If multiple values are specified, a logical OR operator is applied for the condition.
```

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - The Amazon Resource Name (ARN) of the rule.
+* `id` - The Amazon Resource Name (ARN) of the rule.
```
?

PixarV replied:
x

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `health_check` - (Optional, Editable) The health check configuration.
+  The structure of this block is [described below](#health_check).
+* `name` - (Optional) The name of the target group.
+    * _Length of the value_: From 1 to 32 symbols
```
см. анал. ком.

PixarV replied:
Теперь `Value length`

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `name_prefix` - (Optional) Creates a unique name beginning with the specified prefix.
+    * _Length of the value_: From 1 to 6 symbols
+    * _Constraints_:
+        * `name_prefix` cannot be specified with `name`
```
см анал. комм.

PixarV replied:
Поправила формулировку

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+        * The value can contain only Latin letters, numbers and hyphens (`-`)
+        * The value must start and end with a Latin letter or number
+* `name_prefix` - (Optional) Creates a unique name beginning with the specified prefix.
+    * _Length of the value_: From 1 to 6 symbols
```
см. анал. комм.

PixarV replied:
Теперь `Value length`

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+        * `name_prefix` cannot be specified with `name`
+        * The value constraints are the same as `name`
+
+-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with a prefix `tf-`.
```
```suggestion
-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with the prefix `tf-`.
```

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with a prefix `tf-`.
+
+* `protocol_version` - (Optional) The version of HTTP protocol.
```
а тут не надо добавить, что когда протокол выбран HTTP ?

PixarV replied:
Провайдер проигнорирует это поле, если протокол будет не http, но для ясности можно добавить констрейнт

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `protocol` - (Optional, Editable) The protocol used to perform health checks on targets.
+    * _Valid values_: `TCP`, `UDP`, `HTTP`
+    * _Constraints_: `health_check.protocol` and target group `protocol` must be the same
+* `timeout` - (Optional, Editable) The amount of time, in seconds, during which no response means a failed health check.
```
см. анал. комм.

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Constraints_: `health_check.protocol` and target group `protocol` must be the same
+* `timeout` - (Optional, Editable) The amount of time, in seconds, during which no response means a failed health check.
+    * _Valid values_: From 2 to 120
+    * _Constraints_: You cannot set a custom timout value for TCP and UDP target groups (default is 10).
```
```suggestion
    * _Constraints_: You cannot set a custom timeout value for TCP and UDP target groups (default is 10).
```

DiGa66 replied:
может написать, что можно задать только для http (?)

PixarV replied:
Тут еще есть важная информация про дефолтное значение для таких групп (10). Т.е. для http групп можно задать любое, а для tcp и udp - только 10

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
 
-~> **Note** Currently, an NLB (i.e., protocol of `HTTP` or `HTTPS`) can have an invalid `stickiness` block with `type` set to `lb_cookie` as long as `enabled` is set to `false`. However, please update your configurations to avoid errors in a future version of the provider: either remove the invalid `stickiness` block or set the `type` to `source_ip`.
+* `arn` - The Amazon Resource Name (ARN) of the target group.
+* `health_check`:
```
а это не блок?

PixarV replied:
Переоформила как блок

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,130 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `health_check`:
+    * `matcher` - The HTTP code used to check the target availability.
+    * `path` - The destination for the health check request.
+* `id` - The Amazon Resource Name (ARN) of the target group.
```
?

PixarV replied:
x

DiGa66 commented on website/docs/r/lb_target_group_attachment.md:
```diff
@@ -3,77 +3,91 @@ subcategory: "ELB (Elastic Load Balancing)"
-* `port` - (Optional) The port on which targets receive traffic.
+* `target_group_arn` - (Required) The Amazon Resource Name (ARN) of the target group.
+    * _ARN Format_: `arn:c2:elasticloadbalancing::<project-name>@<customer-name>:targetgroup/tg-12345678`
+* `target_id` (Required) The ID of the target.
```
```suggestion
* `target_id` - (Required) The ID of the target.
```

DiGa66 reviewed (changes requested):
остановимся на круглом числе)

DiGa66 commented on website/docs/d/lb_target_group.md:
```diff
@@ -3,48 +3,65 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `id` - The Amazon Resource Name (ARN) of the target group.
+* `port` - The port on which targets receive requests.
+* `protocol` - The protocol that is used for routing traffic to the targets.
+* `protocol_version` - The version of HTTP protocol.
```
т.е. предыдущим параметром выбираем протокол и если это HTTP, то указываем версию?
может тогда так:
```suggestion
* `protocol_version` - The version of the protocol (only for HTTP).
```

PixarV replied:
Это датасорс, тут `protocol_version` - это атрибут, значение которого мы получаем, а не указываем. Поменяла в ресурсе, удалил констрейнт с дублирующейся информацией

DiGa66 commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Value length_: From 1 to 6 symbols
+    * _Constraints_:
+        * `name_prefix` cannot be specified if `name` is set
+        * The value constraints are the same as `name`
```
```suggestion
        * The value constraints are the same as for `name`
```
и в др. местах, если есть

DiGa66 commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Valid values_: From 1 to 65535
+* `protocol` - (Required, Editable) The protocol for a client connection to the load balancer.
+    * _Valid values_:
+        * for network load balancers: `TCP`, `UDP`
```
здесь с прописных (в constraints пункты написаны с прописных)
либо тогда в constraints пункты (если их несколько) со строчных

PixarV replied:
Указала с прописных

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+## Import
+
+Listener rule can be imported using `arn`, e.g.,
```
```suggestion
The listener rule can be imported using `arn`, e.g.,
```
в других местах тогда тоже

DiGa66 commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+* `host` - (Optional, Editable) The name of the host. This component is not percent-encoded.
+    * _Default value_: `#{host}`
+    * _Value length_: From 3 to 255 symbols
+    * _Constraints_: See all the constraints in [ELB API documentation][redirect-action-config]
```
```suggestion
    * _Constraints_: See all the constraints in the [ELB API documentation][redirect-action-config]
```
пишем in the user documentation, тогда здесь тоже;
в других местах тоже поправить

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,137 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Constraints_: You cannot set a custom timeout value for TCP and UDP target groups (default is 10).
+* `unhealthy_threshold` - (Optional, Editable) Number of consecutive failed health checks after which the target status changes to unhealthy.
+    * _Valid values_: From 2 to 10
+    * _Constraints_: For TCP and UDP target groups `healthy_threshold` and `unhealthy_threshold` must be the same
```
```suggestion
    * _Constraints_: For TCP and UDP target groups, `healthy_threshold` and `unhealthy_threshold` must be the same
```

DiGa66 commented on website/docs/r/lb_target_group.md:
```diff
@@ -3,138 +3,137 @@ subcategory: "ELB (Elastic Load Balancing)"
+    * _Constraints_: `health_check.protocol` and target group `protocol` must be the same
+* `timeout` - (Optional, Editable) The amount of time, in seconds, after which no response indicates a failed health check.
+    * _Valid values_: From 2 to 120
+    * _Constraints_: You cannot set a custom timeout value for TCP and UDP target groups (default is 10).
```
лучше в позитиве: 
```suggestion
    * _Constraints_: You can set a custom timeout value for HTTP target groups only (default is 10).
```

cactusFriday commented on internal/service/elbv2/listener_rule.go:
```diff
@@ -81,15 +84,17 @@ func ResourceListenerRule() *schema.Resource {
+							DiffSuppressFunc: suppressIfActionTypeNot(elbv2.ActionTypeEnumForward),
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"target_group": {
```
у нас можно указать 1 таргет группу, мб поправим тоже на MinItems: 1?
```
MinItems: 2,
MaxItems: 5,
```

cactusFriday replied:
Еще Weight у нас 1..256

PixarV replied:
Поправила и MinItems, и Weight

cactusFriday commented on internal/service/elbv2/load_balancer.go:
```diff
@@ -100,18 +99,20 @@ func ResourceLoadBalancer() *schema.Resource {
+				Type:         schema.TypeSet,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
```
```suggestion
```
давай попробуем убрать ForceNew. API позволяет засетить как subnet_mappings, так и subnets. Как я понимаю, в коде resourceUpdate нужно будет расширить ифчик `if d.HasChange("subnets") && !d.IsNewResource() {`

PixarV replied:
Добавила возможность обновлять список блоков `subnet_mapping`

cactusFriday commented on website/docs/d/lb.md:
```diff
@@ -3,49 +3,60 @@ subcategory: "ELB (Elastic Load Balancing)"
-* `tags` - (Optional) A mapping of tags, each pair of which must exactly match a pair on the desired load balancer.
+* `arn` - (Optional) The Amazon Resource Name (ARN) of the load balancer.
+    * _ARN Format_: `arn:c2:elasticloadbalancing::<project-name>@<customer-name>:loadbalancer/<app|net>/lb-12345678`
+* `name` - (Optional) The name of the load balancer.
```
```suggestion
* `name` - The name of the load balancer.
```
name у нас required, посмотри еще в схему балансера, пожалуйста

PixarV replied:
**Для датасорса aws_load_balancer**: для поиска балансера провайдер позволяет указать либо arn, либо имя балансера, либо теги балансера. Так что поле `name` тут оправдано опциональное

**Для ресурса aws_load_balancer**: провайдер позволяет указать либо `name`, либо `name_prefix` (провайдер сгенерит имя с таким префиксом), либо вообще ничего не указать и тогда провайдер сам сгенерит имя с префиксом `tf-lb`. Эта схема работы перетекла от aws. Т.е. на уровне провайдера это все таки опциональное поле

cactusFriday replied:
а блин, я дурилка картонная, это датасорс. Прошу прощения

cactusFriday commented on website/docs/d/lb_target_group.md:
```diff
@@ -3,48 +3,65 @@ subcategory: "ELB (Elastic Load Balancing)"
-* `name` - (Optional) The unique name of the target group.
+* `arn` - (Optional) The Amazon Resource Name (ARN) of the target group.
+    * _ARN Format_: `arn:c2:elasticloadbalancing::<project-name>@<customer-name>:targetgroup/tg-12345678`
+* `name` - (Optional) The name of the target group.
```
tg name тоже required

PixarV replied:
Тут такая же схема работы как у балансера.

cactusFriday commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+
+-> **Note** If `name` and `name_prefix` are not specified, Terraform will autogenerate a name with the prefix `tf-lb`.
+
+* `subnet_mapping` - (Optional) List of subnet-ID-to-IP-address mappings.
```
маппинги тоже можно изменять
(Optionsl. Editable)

cactusFriday commented on website/docs/r/lb.md:
```diff
@@ -3,169 +3,171 @@ subcategory: "ELB (Elastic Load Balancing)"
+        * The `subnets` argument is required if `subnet_mapping` is not specified
+        * All subnets must be from different availability zones
+
+~> **Note** You can only add new subnets to the `subnets` list, subnets cannot be removed.
```
or subnet_mappings

cactusFriday commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
+The `default_action` block has the following structure:
 
-#### authenticate_cognito
+* `type` - (Required, Editable) The type of the routing action.
+    * _Valid values_: `forward`
```
nit: 
тут не понятно. type - editable, но valid-values - 1 значение. Может убрать editable?

PixarV replied:
Я бы оставила, формально это все таки редактируемое поле, тк оно не ForceNew. Тем более здесь потенциально появятся другие значения и можно будет забыть докинуть признак Editable)

cactusFriday commented on website/docs/r/lb_listener.md:
```diff
@@ -3,367 +3,156 @@ subcategory: "ELB (Elastic Load Balancing)"
+* `arn` - (Required, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic to.
+    * _ARN Format_: `arn:c2:elasticloadbalancing::<project-name>@<customer-name>:targetgroup/tg-12345678`
+* `weight` - (Optional, Editable) The weight of the target group.
+    * _Valid values_: From 0 to 999
```
```suggestion
    * _Valid values_: From 0 to 256
```

cactusFriday commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+* `status_code` - (Required, Editable) The HTTP code of the response.
+    * _Valid values_: `2XX`, `4XX`, `5XX`, where `X` is a digit
+* `message_body` - (Optional, Editable) The message of the response.
+    * _Value length_: From 0 to 1024 symbols
```
nit:
можно дописать, что если status_code = 204, то message_body указывать в request нельзя. Такое валидируется.

cactusFriday commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+
+* `target_group` - (Required, Editable) List of target groups to forward traffic to.
+  The structure of this block is [described below](#target_group).
+    * _List size_: From 2 to 5 elements
```
```suggestion
    * _List size_: From 1 to 5 elements
```

cactusFriday commented on website/docs/r/lb_listener_rule.md:
```diff
@@ -0,0 +1,353 @@
+* `arn` - (Required, Editable) The Amazon Resource Name (ARN) of the target group to forward traffic to.
+    * _ARN Format_: `arn:c2:elasticloadbalancing::<project-name>@<customer-name>:targetgroup/tg-12345678`
+* `weight` - (Optional, Editable) The weight of the target group.
+    * _Valid values_: From 0 to 999
```
```suggestion
    * _Valid values_: From 0 to 256
```

cactusFriday commented on website/docs/r/lb_target_group_attachment.md:
```diff
@@ -3,77 +3,90 @@ subcategory: "ELB (Elastic Load Balancing)"
+resource "aws_lb_target_group_attachment" "example" {
+  target_group_arn = aws_lb_target_group.example.arn
+  target_id        = aws_instance.example.id
+  port             = 81
```
давай либо уберем, либо 80 оставим. Возможность менять порт конечно есть, но лучше не показывать это пользователю, а то он начнет неправильно сервисом пользоваться. Пусть будет порт таргет группы выше.

PixarV replied:
Убрала порт из примера.

Commit:
elbv2,website: mark default_action.target_group_arn as deprecated

The `target_group_arn` argument has been marked deprecated to remove
one of the ways to specify a target group to forward traffic to.
The `forward` block is now the preferred way to do it.

In case of one target group, it doesn't matter what arguments are
specified in the configuration, the ELB API will populate both
the `TargetGroupArn` and `ForwardConfig` parameters in the response.
The provider can't save both parameters to the state because there
would be a difference between the state and the configuration (where
only one argument is specified). So now `target_group_arn` is set from
the response only if it was specified in the configuration. Otherwise,
`forward` is set (also on import).

Commit:
elbv2,website: listener rule: remove action.target_group_arn

Now there is only one way to specify a target group to forward traffic
to.

Commit:
elbv2: retry target deregistering if OperationInProgress, refactoring
